x-pack/plugins/security/server/lib/watch_status_and_license_to_initialize.test.js

```diff
@@ -0,0 +1,221 @@
+    const initializeMock = jest.fn();
+
+    watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+
+    expect(initializeMock).not.toHaveBeenCalled();
+    expect(downstreamPlugin.status[state]).toHaveBeenCalledTimes(1);
+    expect(downstreamPlugin.status[state]).toHaveBeenCalledWith(message);
+  });
+});
+
+test(`calls initialize and doesn't immediately set downstream status when the initial status is green`, () => {
+  const pluginId = 'foo-plugin';
+  const { mockXpackMainPlugin, mockFeature } = createMockXpackMainPluginAndFeature(pluginId);
+  mockXpackMainPlugin.mock.setStatus('green', 'green is now the state');
+  const licenseCheckResults = Symbol();
+  mockFeature.mock.setLicenseCheckResults(licenseCheckResults);
+  const downstreamPlugin = createMockDownstreamPlugin(pluginId);
+  const initializeMock = jest.fn().mockImplementation(() => new Promise(() => {}));
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+
+  expect(initializeMock).toHaveBeenCalledTimes(1);
+  expect(initializeMock).toHaveBeenCalledWith(licenseCheckResults);
+  expect(downstreamPlugin.status.green).toHaveBeenCalledTimes(0);
+});
+
+test(`sets downstream plugin's status to green when initialize resolves`, (done) => {
+  const pluginId = 'foo-plugin';
+  const { mockXpackMainPlugin, mockFeature } = createMockXpackMainPluginAndFeature(pluginId);
+  mockXpackMainPlugin.mock.setStatus('green', 'green is now the state');
+  const licenseCheckResults = Symbol();
+  mockFeature.mock.setLicenseCheckResults(licenseCheckResults);
+  const downstreamPlugin = createMockDownstreamPlugin(pluginId);
+  const initializeMock = jest.fn().mockImplementation(() => Promise.resolve());
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+
+  expect(initializeMock).toHaveBeenCalledTimes(1);
+  expect(initializeMock).toHaveBeenCalledWith(licenseCheckResults);
+  downstreamPlugin.status.green.mockImplementation(actualMessage =>
+  {
+    expect(actualMessage).toBe('Ready');
+    done();
+  });
+});
+
+test(`sets downstream plugin's status to red when initialize rejects`, (done) => {
+  const pluginId = 'foo-plugin';
+  const errorMessage = 'the error message';
+  const { mockXpackMainPlugin, mockFeature } = createMockXpackMainPluginAndFeature(pluginId);
+  mockXpackMainPlugin.mock.setStatus('green');
+  const licenseCheckResults = Symbol();
+  mockFeature.mock.setLicenseCheckResults(licenseCheckResults);
+  const downstreamPlugin = createMockDownstreamPlugin(pluginId);
+  const initializeMock = jest.fn().mockImplementation(() => Promise.reject(new Error(errorMessage)));
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+
+  expect(initializeMock).toHaveBeenCalledTimes(1);
+  expect(initializeMock).toHaveBeenCalledWith(licenseCheckResults);
+  downstreamPlugin.status.red.mockImplementation(message => {
```

Comment:
do you think there's benefit to checking that `green`, `yellow`, etc aren't called?

Reply:
Yeah, we can make the mocks throw errors by default to make sure they aren't accidentally being called.